### DISCUSSION TITLE
v1.2.0

### DISCUSSION THREAD

**Discussion by ggerganov** [ggml-org/whisper.cpp#467] [Announcements]
## Overview

In this release we significantly reduce the memory usage during inference by introducing "scratch" buffers to `ggml`.
The new memory requirements per model are as follows:

| Model  | Disk   | Mem (Old)    | Mem (New)                                        |
| ---    | ---    | ---     | ---                                        |
| tiny   |  75 MB | ~390 MB | ~125 MB |
| base   | 142 MB | ~500 MB | ~210 MB |
| small  | 466 MB | ~1.0 GB | ~600 MB |
| medium | 1.5 GB | ~2.6 GB | ~1.7 GB |
| large  | 2.9 GB | ~4.7 GB | ~3.3 GB |

It's a simple idea that instead of creating a new memory buffer for each new tensor in the computation, we reuse the memory of old tensors that are no longer needed. The implementation is in PR #431. It's not very clean - I think there is some better way to do this, but for now it will work.

Additionally, there might be some inference speed improvements on Apple Silicon in the Decoder part of the transformer. I haven't done proper benchmarks, but seems there is about ~30% performance boost. The results are identical to `v1.1.1`.

#### Core `ggml` / `whisper`
* `whisper` : PPC64 big-endian support by @fitzsim in https://github.com/ggerganov/whisper.cpp/pull/398
* `whisper` : condition sampled timestamp tokens to be monotonically increasing by @ggerganov in https://github.com/ggerganov/whisper.cpp/pull/425
* `wasm` : fix typo in helper.js by @bhbs in https://github.com/ggerganov/whisper.cpp/pull/459
* `ggml`/`whisper` : reduce memory usage during inference by @ggerganov in https://github.com/ggerganov/whisper.cpp/pull/431

#### Bindings
* `ci` : run workflows on pull requests + bindings depend on .h by @ggerganov in https://github.com/ggerganov/whisper.cpp/pull/446
* `go` : added wrappers to reset and print timings by @glaslos in https://github.com/ggerganov/whisper.cpp/pull/436
* `go` : add WhisperLangAutoDetect method to go binding by @RobinXL in https://github.com/ggerganov/whisper.cpp/pull/451
* `go` : add wrapper for system info by @glaslos in https://github.com/ggerganov/whisper.cpp/pull/456
* `go` : support "auto" as an option when set language by @polarmoon in https://github.com/ggerganov/whisper.cpp/pull/462

#### Examples
* `whisper.wasm` : add labels for easier radio selection by @kokes in https://github.com/ggerganov/whisper.cpp/pull/435
* `livestream.sh` : run main with model arg instead of default by @EricTendian in https://github.com/ggerganov/whisper.cpp/pull/453
* `main` : CSV format export trimmed spaces fix by @alex-bacart in https://github.com/ggerganov/whisper.cpp/pull/444
* `addon.node` : using whisper as a Node.js addon by @chenqianhe in https://github.com/ggerganov/whisper.cpp/pull/443

## New Contributors
* @kokes made their first contribution in https://github.com/ggerganov/whisper.cpp/pull/435
* @glaslos made their first contribution in https://github.com/ggerganov/whisper.cpp/pull/436
* @EricTendian made their first contribution in https://github.com/ggerganov/whisper.cpp/pull/453
* @RobinXL made their first contribution in https://github.com/ggerganov/whisper.cpp/pull/451
* @alex-bacart made their first contribution in https://github.com/ggerganov/whisper.cpp/pull/444
* @bhbs made their first contribution in https://github.com/ggerganov/whisper.cpp/pull/459
* @polarmoon made their first contribution in https://github.com/ggerganov/whisper.cpp/pull/462
* @chenqianhe made their first contribution in https://github.com/ggerganov/whisper.cpp/pull/443

**Full Changelog**: https://github.com/ggerganov/whisper.cpp/compare/v1.1.1...v1.2.0

## Highlights

I'll use these release notes to write some random thoughts about the project - sort of a short blog post.

I'm really happy with how `whisper.cpp` turned out to be so far. There is a very positive reception in the ML community - most people seem to be excited by the simplicity of the implementation and the fact that it is quite self-contained. I receive a lot of questions about the project and about various ideas that it can be applied to. I really enjoy it and I try to respond to everyone!

I also find it very satisfying that there are so many contributions already happening by so many people. To me this illustrates the power of open-source collaboration. The contributions not only improve the functionality and the quality of the code, but also help to generate various new ideas and approaches to explore.

Another interesting thing is that the project keeps on giving. Every time I start to think that now is a good time to put it in the background for a while and focus on other stuff, some new cool idea pops up and I can't help but start working on it. Having this custom implementation allows me to interact with the model on a lower level which opens some interesting ways to explore it.

So far the development has been focused on improving the performance, expanding the platform coverage and having robust decoding strategies with a variety of examples. During this time, there have been several ideas that accumulated over-time which I find interesting to explore (diarization, token-level timestamps, improved timestamp accuracy, etc). I think I'll try to focus more on these in the future and see if I can achieve something interesting.

---

- **Windows port of `whisper.cpp` utilising vendor-agnostic GPGPU based on DirectCompute by @Const-me**

    https://github.com/Const-me/Whisper

---

- **"The New Yorker" article featuring `whisper.cpp`**

<div align="center">
<h2><a href="https://www.newyorker.com/tech/annals-of-technology/whispers-of-ais-modular-future">Whispers of A.I.’s Modular Future</a></h2>
<a href="https://www.newyorker.com/tech/annals-of-technology/whispers-of-ais-modular-future"><img width="1663" alt="image" src="https://media.newyorker.com/photos/63d93e688b2aff35d30ef8e2/master/w_2560,c_limit/Somers_final.jpg"></a>
</div>

<hr /><em>This discussion was created from the release <a href='https://github.com/ggerganov/whisper.cpp/releases/tag/v1.2.0'>v1.2.0</a>.</em>

**Comment by jerodsanto** [comment]
Congrats on another awesome release! I'm _super_ impressed by this project @ggerganov (and contributors). Two thoughts:

1. We would _love_ to have you join us on [The Changelog](https://changelog.fm) podcast to discuss your work!
2.  Diarization would be a huge/standout feature IMO, since it's not supported at all by OpenAI's Whisper model

I am happy to provide more info to you about the podcast, if you're curious. Here works or you can email me jerod@changelog.com if you'd like to discuss privately. 💚

**Reply by ggerganov** [reply]
@jerodsanto Let's give it a try 😄 

For me best time for chat is in my evening: 7-9pm UTC+2

**Reply by jerodsanto** [reply]
Exciting! Our normal recording day is Wednesday. How about [March 15th at 11am our time/7pm your time](https://time.is/1678896000)?

**Reply by ggerganov** [reply]
Yes, this works for me!

**Reply by strangelearning** [reply]
Please post the link once live @jerodsanto !

**Reply by jerodsanto** [reply]
Will do @strangelearning!

@ggerganov you should have a calendar invite in your inbox. Also check out our guest guide and holler back any questions you have. 

https://changelog.fm/guest

**Reply by ggerganov** [reply]
@strangelearning 

Here's the link to the episode:

<audio data-theme="night" data-src="https://changelog.com/podcast/532/embed" src="https://op3.dev/e/https://cdn.changelog.com/uploads/podcast/532/the-changelog-532.mp3" preload="none" class="changelog-episode" controls></audio><p><a href="https://changelog.com/podcast/532">The Changelog 532: Bringing Whisper and LLaMA to the masses</a> – Listen on <a href="https://changelog.com/">Changelog.com</a></p>

**Comment by ggerganov** [comment]
Giving more visibility to this - there has been a bug for a while that can be triggered if using the language auto-detect feature:

https://github.com/ggerganov/whisper.cpp/pull/520

It's now fixed on latest `master`, but the issue exists in all released version and can cause random crashes.

**Comment by sanchitram1** [comment]
I've not had any issues with it so far, and absolutely loved playing around with the model.  I am packaging it up to be available via [tea](https://github.com/teaxyz/cli), so the following commands will work for any user who has tea:

```zsh
$ whisper.main
$ whisper.stream
$ whisper.command
```

I'll let you know when it's available!
